# [R]apify
### The following program will take a CSV of spotify streaming history, and use it to determine the most popular artists and songs. 

NOTE: All "install.packages" lines will be commented out, if at any point something is not running due to an "object does not exist" error, uncomment and install the required package. 

In [1]:
# confirm that the file exists
file.exists("spotify.csv")

[1] TRUE

In [2]:
# creating a data frame using the csv
streaming_history <- read.csv(file = "spotify.csv")

# viewing the first couple rows to ensure everything looks correct
str(streaming_history)

'data.frame':	157779 obs. of  11 variables:
 $ ts          : chr  "2016-02-20T03:28:50Z" "2016-02-20T03:30:36Z" "2016-02-20T03:30:38Z" "2016-02-20T03:30:57Z" ...
 $ ms_played   : int  20200 96896 2530 18459 14466 1648 199593 178600 213093 7589 ...
 $ song_title  : chr  "Gasoline" "Gasoline" "Gasoline" "Gasoline" ...
 $ artist_name : chr  "Halsey" "Halsey" "Halsey" "Halsey" ...
 $ album_name  : chr  "BADLANDS" "BADLANDS" "BADLANDS" "BADLANDS" ...
 $ track_url   : chr  "spotify:track:2IO7yf562c1zLzpanal1DT" "spotify:track:2IO7yf562c1zLzpanal1DT" "spotify:track:2IO7yf562c1zLzpanal1DT" "spotify:track:2IO7yf562c1zLzpanal1DT" ...
 $ reason_start: chr  "trackdone" "clickrow" "clickrow" "clickrow" ...
 $ reason_end  : chr  "endplay" "endplay" "endplay" "endplay" ...
 $ shuffle     : logi  FALSE FALSE FALSE FALSE FALSE FALSE ...
 $ skipped     : logi  TRUE TRUE TRUE TRUE TRUE TRUE ...
 $ offline     : logi  FALSE FALSE FALSE FALSE FALSE FALSE ...


In [3]:
# summary can tell us a bit about each variable.
summary(streaming_history)

      ts              ms_played        song_title        artist_name       
 Length:157779      Min.   :      0   Length:157779      Length:157779     
 Class :character   1st Qu.:   1438   Class :character   Class :character  
 Mode  :character   Median :   6243   Mode  :character   Mode  :character  
                    Mean   :  78671                                        
                    3rd Qu.: 183808                                        
                    Max.   :3727766                                        
  album_name         track_url         reason_start        reason_end       
 Length:157779      Length:157779      Length:157779      Length:157779     
 Class :character   Class :character   Class :character   Class :character  
 Mode  :character   Mode  :character   Mode  :character   Mode  :character  
                                                                            
                                                                            
      

## Looks like some cleaning is required
Did you notice that ms_played has a min of 0? I am not interested in counting songs with 0 ms_played as individual listens. Lets remove those. 

I also have a max ms_played of 3727766 which roughly translates to.... 62 minutes? That's not right. I'll be setting a listening cap of 15 minutes to account for some of the extra long songs I listen to. Looking at you, Guns n' Roses. 

I am also going to remove any songs where I stopped listening after 20000 ms, since that means I only listened for about 20 seconds.

In [4]:
# if any songs play for less than 2000 ms or over 900000 ms (15 mins) mark them NA
streaming_history$ms_played <- ifelse(streaming_history$ms_played > 20000 & streaming_history$ms_played < 900000, streaming_history$ms_played, NA)
# remove null values
cleaned_streaming_history <- na.omit(streaming_history)
summary(streaming_history)
summary(cleaned_streaming_history)

      ts              ms_played       song_title        artist_name       
 Length:157779      Min.   : 20002   Length:157779      Length:157779     
 Class :character   1st Qu.:127349   Class :character   Class :character  
 Mode  :character   Median :201053   Mode  :character   Mode  :character  
                    Mean   :186739                                        
                    3rd Qu.:237706                                        
                    Max.   :884329                                        
                    NA's   :93256                                         
  album_name         track_url         reason_start        reason_end       
 Length:157779      Length:157779      Length:157779      Length:157779     
 Class :character   Class :character   Class :character   Class :character  
 Mode  :character   Mode  :character   Mode  :character   Mode  :character  
                                                                            
               

      ts              ms_played       song_title        artist_name       
 Length:64523       Min.   : 20002   Length:64523       Length:64523      
 Class :character   1st Qu.:127349   Class :character   Class :character  
 Mode  :character   Median :201053   Mode  :character   Mode  :character  
                    Mean   :186739                                        
                    3rd Qu.:237706                                        
                    Max.   :884329                                        
  album_name         track_url         reason_start        reason_end       
 Length:64523       Length:64523       Length:64523       Length:64523      
 Class :character   Class :character   Class :character   Class :character  
 Mode  :character   Mode  :character   Mode  :character   Mode  :character  
                                                                            
                                                                            
             

## As we can see...
The length of all my variables dropped from 157779 in the first summary to 64523 in the second. The size of my dataset has decreased significantly. Normally this is a bad sign but I knew going into this that this dataset was going to be very inconsistent (since its based on the whims of my own listening habits) so for this instance, its okay. Lets move on. 

In [5]:
#install.packages("data.table")
library(data.table)

# converting streaming history into a data table
dt_streaming = data.table(cleaned_streaming_history)

In [6]:
# making tables for each listening year
listening_2016 <- dt_streaming[
  ts >= ("2016-01-01T00:00:00Z") &
  ts <= ("2016-12-31T23:59:59Z")
]

listening_2017 <- dt_streaming[
  ts >= ("2017-01-01T00:00:00Z") &
  ts <= ("2017-12-31T23:59:59Z")
]

listening_2018 <- dt_streaming[
  ts >= ("2018-01-01T00:00:00Z") &
  ts <= ("2018-12-31T23:59:59Z")
]

listening_2019 <- dt_streaming[
  ts >= ("2019-01-01T00:00:00Z") &
  ts <= ("2019-12-31T23:59:59Z")
]

listening_2020 <- dt_streaming[
  ts >= ("2020-01-01T00:00:00Z") &
  ts <= ("2020-12-31T23:59:59Z")
]

listening_2021 <- dt_streaming[
  ts >= ("2021-01-01T00:00:00Z") &
  ts <= ("2021-12-31T23:59:59Z")
]

listening_2022 <- dt_streaming[
  ts >= ("2022-01-01T00:00:00Z") &
  ts <= ("2022-12-31T23:59:59Z")
]

listening_2023 <- dt_streaming[
  ts >= ("2023-01-01T00:00:00Z") &
  ts <= ("2023-12-31T23:59:59Z")
]

listening_2024 <- dt_streaming[
  ts >= ("2024-01-01T00:00:00Z") &
  ts <= ("2024-12-31T23:59:59Z")
]

listening_2025 <- dt_streaming[
  ts >= ("2025-01-01T00:00:00Z") &
  ts <= ("2025-12-31T23:59:59Z")
]

# Streaming habits for all time

In [7]:
# the total song counts for all time listening
# confirm that the song name and artist are the same to ensure its a unique song and sort by descending order
# install.packages("dplyr")
library(dplyr)
total_song_counts <- count(dt_streaming, song_title, artist_name, sort = TRUE)

# knitr library so output can be pretty
# install.packages("knitr")
library(knitr)
kable(head(total_song_counts, 20), col.names = c("Song", "Artist", "Plays"), caption = "Top 20 songs of all time")

# the total artist counts for all time listening
# sort by descending order and use knitr library so output can be pretty
total_artist_counts <- count(dt_streaming, artist_name, sort = TRUE)
kable(head(total_artist_counts, 10), col.names = c("Artist", "Plays"), caption = "Top 10 artists of all time")


Attaching package: 'dplyr'


The following objects are masked from 'package:data.table':

    between, first, last


The following objects are masked from 'package:stats':

    filter, lag


The following objects are masked from 'package:base':

    intersect, setdiff, setequal, union






Table: Top 20 songs of all time

|Song                                          |Artist              | Plays|
|:---------------------------------------------|:-------------------|-----:|
|Stand By You                                  |Rachel Platten      |   171|
|The Sound of Silence                          |Disturbed           |   137|
|Chocolate                                     |The 1975            |   130|
|Starlight                                     |Muse                |   126|
|Revolt                                        |Muse                |   123|
|The Only Hope for Me Is You                   |My Chemical Romance |   114|
|Undertale (Music Box)                         |Jayn                |   114|
|United States of Eurasia (+Collateral Damage) |Muse                |   113|
|Human                                         |The Killers         |   112|
|Monkey Tree - UK Mix                          |Mother Mother       |   112|
|High School                             



Table: Top 10 artists of all time

|Artist              | Plays|
|:-------------------|-----:|
|Muse                |  2973|
|Twenty One Pilots   |  2153|
|One Direction       |  1746|
|Panic! At The Disco |  1444|
|My Chemical Romance |  1439|
|Imagine Dragons     |  1414|
|Ghost               |  1365|
|Bleachers           |   947|
|Halsey              |   881|
|Lorde               |   809|

# Streaming habits by year

In [8]:
total_song_counts_16 <- count(listening_2016, song_title, artist_name, sort = TRUE)
total_artist_counts_16 <- count(listening_2016, artist_name, sort = TRUE)
total_song_counts_17 <- count(listening_2017, song_title, artist_name, sort = TRUE)
total_artist_counts_17 <- count(listening_2017, artist_name, sort = TRUE)
total_song_counts_18 <- count(listening_2018, song_title, artist_name, sort = TRUE)
total_artist_counts_18 <- count(listening_2018, artist_name, sort = TRUE)
total_song_counts_19 <- count(listening_2019, song_title, artist_name, sort = TRUE)
total_artist_counts_19 <- count(listening_2019, artist_name, sort = TRUE)
total_song_counts_20 <- count(listening_2020, song_title, artist_name, sort = TRUE)
total_artist_counts_20 <- count(listening_2020, artist_name, sort = TRUE)
total_song_counts_21 <- count(listening_2021, song_title, artist_name, sort = TRUE)
total_artist_counts_21 <- count(listening_2021, artist_name, sort = TRUE)
total_song_counts_22 <- count(listening_2022, song_title, artist_name, sort = TRUE)
total_artist_counts_22 <- count(listening_2022, artist_name, sort = TRUE)
total_song_counts_23 <- count(listening_2023, song_title, artist_name, sort = TRUE)
total_artist_counts_23<- count(listening_2023, artist_name, sort = TRUE)
total_song_counts_24 <- count(listening_2024, song_title, artist_name, sort = TRUE)
total_artist_counts_24 <- count(listening_2024, artist_name, sort = TRUE)
total_song_counts_25 <- count(listening_2025, song_title, artist_name, sort = TRUE)
total_artist_counts_25 <- count(listening_2025, artist_name, sort = TRUE)

In [9]:
kable(head(total_song_counts_16, 10), col.names = c("Song", "Artist", "Plays"), caption = "Top 10 songs of 2016")
kable(head(total_artist_counts_16, 5), col.names = c("Artist", "Plays"), caption = "Top 5 artists of 2016")
kable(head(total_song_counts_17, 10), col.names = c("Song", "Artist", "Plays"), caption = "Top 10 songs of 2017")
kable(head(total_artist_counts_17, 5), col.names = c("Artist", "Plays"), caption = "Top 5 artists of 2017")
kable(head(total_song_counts_18, 10), col.names = c("Song", "Artist", "Plays"), caption = "Top 10 songs of 2018")
kable(head(total_artist_counts_18, 5), col.names = c("Artist", "Plays"), caption = "Top 5 artists of 2018")
kable(head(total_song_counts_19, 10), col.names = c("Song", "Artist", "Plays"), caption = "Top 10 songs of 2019")
kable(head(total_artist_counts_19, 5), col.names = c("Artist", "Plays"), caption = "Top 5 artists of 2019")
kable(head(total_song_counts_20, 10), col.names = c("Song", "Artist", "Plays"), caption = "Top 10 songs of 2020")
kable(head(total_artist_counts_20, 5), col.names = c("Artist", "Plays"), caption = "Top 5 artists of 2020")
kable(head(total_song_counts_21, 10), col.names = c("Song", "Artist", "Plays"), caption = "Top 10 songs of 2021")
kable(head(total_artist_counts_21, 5), col.names = c("Artist", "Plays"), caption = "Top 5 artists of 2021")
kable(head(total_song_counts_22, 10), col.names = c("Song", "Artist", "Plays"), caption = "Top 10 songs of 2022")
kable(head(total_artist_counts_22, 5), col.names = c("Artist", "Plays"), caption = "Top 5 artists of 2022")
kable(head(total_song_counts_23, 10), col.names = c("Song", "Artist", "Plays"), caption = "Top 10 songs of 2023")
kable(head(total_artist_counts_23, 5), col.names = c("Artist", "Plays"), caption = "Top 5 artists of 2023")
kable(head(total_song_counts_24, 10), col.names = c("Song", "Artist", "Plays"), caption = "Top 10 songs of 2024")
kable(head(total_artist_counts_24, 5), col.names = c("Artist", "Plays"), caption = "Top 5 artists of 2024")
kable(head(total_song_counts_25, 10), col.names = c("Song", "Artist", "Plays"), caption = "Top 10 songs of 2025")
kable(head(total_artist_counts_25, 5), col.names = c("Artist", "Plays"), caption = "Top 5 artists of 2025")



Table: Top 10 songs of 2016

|Song                                                                       |Artist              | Plays|
|:--------------------------------------------------------------------------|:-------------------|-----:|
|Stand By You                                                               |Rachel Platten      |   171|
|The Sound of Silence                                                       |Disturbed           |   100|
|Angel With A Shotgun                                                       |The Cab             |    74|
|Heathens                                                                   |Twenty One Pilots   |    64|
|Undertale (Music Box)                                                      |Jayn                |    64|
|CHARIOTS OF FIRE - Café Buddha Del Mar Bar Mix, as made famous by Vangelis |Eternal-Flame       |    56|
|You're Gonna Go Far, Kid                                                   |The Offspring       |    53|
|High School   



Table: Top 5 artists of 2016

|Artist              | Plays|
|:-------------------|-----:|
|Twenty One Pilots   |   892|
|Panic! At The Disco |   462|
|Imagine Dragons     |   284|
|Halsey              |   216|
|Rachel Platten      |   171|



Table: Top 10 songs of 2017

|Song                                                       |Artist          | Plays|
|:----------------------------------------------------------|:---------------|-----:|
|I Dreamed A Dream (Glee Cast Version) (feat. Idina Menzel) |Glee Cast       |    54|
|You Don't Own Me (feat. G-Eazy)                            |SAYGRACE        |    42|
|Play That Song                                             |Train           |    37|
|I'm Ready                                                  |AJR             |    34|
|True Love                                                  |P!nk            |    34|
|Copacabana (Glee Cast Version)                             |Glee Cast       |    33|
|Feeling Good                                               |Michael Bublé   |    33|
|Can't Help Falling in Love                                 |Haley Reinhart  |    32|
|Hallelujah                                                 |Tori Kelly      |    32|
|Weekend In Monaco     



Table: Top 5 artists of 2017

|Artist              | Plays|
|:-------------------|-----:|
|Imagine Dragons     |   616|
|Twenty One Pilots   |   507|
|Steven Universe     |   406|
|Halsey              |   335|
|Panic! At The Disco |   244|



Table: Top 10 songs of 2018

|Song                                                                                 |Artist              | Plays|
|:------------------------------------------------------------------------------------|:-------------------|-----:|
|Da Ya Think I'm Sexy                                                                 |Rod Stewart         |    43|
|Panic Room                                                                           |Au/Ra               |    41|
|Never Enough                                                                         |Loren Allred        |    40|
|War!                                                                                 |Zach Callison       |    39|
|Everybody Wants To Rule The World                                                    |Tears For Fears     |    37|
|Golden Slumbers / Carry That Weight - From "Sing" Original Motion Picture Soundtrack |Jennifer Hudson     |    36|
|The Love Club                           



Table: Top 5 artists of 2018

|Artist                             | Plays|
|:----------------------------------|-----:|
|Panic! At The Disco                |   386|
|Twenty One Pilots                  |   381|
|Lorde                              |   348|
|Imagine Dragons                    |   236|
|Original Broadway Cast of Hamilton |   232|



Table: Top 10 songs of 2019

|Song                           |Artist              | Plays|
|:------------------------------|:-------------------|-----:|
|Kissaphobic                    |Make Out Monday     |    37|
|Private Fears in Public Places |Front Porch Step    |    37|
|Lethal Combination             |The Wombats         |    34|
|Promise Me                     |Jack & Jack         |    34|
|The Only Hope for Me Is You    |My Chemical Romance |    33|
|Monkey Tree - UK Mix           |Mother Mother       |    32|
|And There You Are              |Alex Walker Smith   |    31|
|Freaks                         |Jordan Clarke       |    31|
|Girls Like Girls               |Hayley Kiyoko       |    31|
|Diana                          |One Direction       |    30|



Table: Top 5 artists of 2019

|Artist            | Plays|
|:-----------------|-----:|
|One Direction     |   636|
|Bleachers         |   235|
|YUNGBLUD          |   197|
|Twenty One Pilots |   172|
|Jack & Jack       |   134|



Table: Top 10 songs of 2020

|Song                                          |Artist         | Plays|
|:---------------------------------------------|:--------------|-----:|
|Therapy                                       |New Politics   |    47|
|Human                                         |The Killers    |    46|
|Breakup Haircut                               |Danny and Alex |    44|
|Sisters                                       |Saint Motel    |    43|
|Starlight                                     |Muse           |    42|
|Girl Crush                                    |New Politics   |    40|
|Goodbye Copenhagen                            |New Politics   |    40|
|United States of Eurasia (+Collateral Damage) |Muse           |    40|
|Cold Cold Man                                 |Saint Motel    |    38|
|Cemetery                                      |COIN           |    37|



Table: Top 5 artists of 2020

|Artist        | Plays|
|:-------------|-----:|
|Muse          |   551|
|One Direction |   326|
|Blue October  |   315|
|New Politics  |   256|
|Bleachers     |   227|



Table: Top 10 songs of 2021

|Song                       |Artist              | Plays|
|:--------------------------|:-------------------|-----:|
|Bullet for Your Sweetheart |Make Out Monday     |    42|
|Potential Breakup Song     |Aly & AJ            |    37|
|Revolt                     |Muse                |    37|
|Bulletproof Heart          |My Chemical Romance |    35|
|Creep                      |Radiohead           |    34|
|On Melancholy Hill         |Gorillaz            |    34|
|Disenchanted               |My Chemical Romance |    33|
|Reapers                    |Muse                |    33|
|Uprising                   |Muse                |    33|
|Ballroom Blitz             |Sweet               |    32|



Table: Top 5 artists of 2021

|Artist              | Plays|
|:-------------------|-----:|
|My Chemical Romance |   447|
|Muse                |   446|
|One Direction       |   225|
|Mother Mother       |   142|
|Neon Trees          |   140|



Table: Top 10 songs of 2022

|Song                                 |Artist            | Plays|
|:------------------------------------|:-----------------|-----:|
|Call Me Little Sunshine              |Ghost             |    45|
|Pressure                             |Muse              |    42|
|Pleasure                             |Betcha            |    40|
|Jenny Was A Friend Of Mine           |The Killers       |    39|
|Revolt                               |Muse              |    38|
|Houdini                              |Foster The People |    37|
|You Make Me Feel Like It’s Halloween |Muse              |    35|
|Deus In Absentia                     |Ghost             |    33|
|Mary On A Cross                      |Ghost             |    33|
|Neon Tiger                           |The Killers       |    33|



Table: Top 5 artists of 2022

|Artist              | Plays|
|:-------------------|-----:|
|Muse                |   798|
|Ghost               |   426|
|One Direction       |   274|
|My Chemical Romance |   264|
|flipturn            |   224|



Table: Top 10 songs of 2023

|Song                              |Artist          | Plays|
|:---------------------------------|:---------------|-----:|
|Thunderstruck                     |AC/DC           |    46|
|Enjoy the Silence - 2006 Remaster |Depeche Mode    |    43|
|Darkness At The Heart Of My Love  |Ghost           |    42|
|Plug in Baby                      |Muse            |    35|
|Faith                             |Ghost           |    33|
|Cirice                            |Ghost           |    30|
|Cliffs Of Dover - Instrumental    |Eric Johnson    |    30|
|My Girlfriend's Girlfriend        |Type O Negative |    29|
|You Shook Me All Night Long       |AC/DC           |    29|
|1983                              |Neon Trees      |    27|



Table: Top 5 artists of 2023

|Artist              | Plays|
|:-------------------|-----:|
|Ghost               |   434|
|Muse                |   428|
|Neon Trees          |   180|
|The Killers         |   133|
|My Chemical Romance |   110|



Table: Top 10 songs of 2024

|Song                                   |Artist           | Plays|
|:--------------------------------------|:----------------|-----:|
|Silver                                 |Nosferatu        |    45|
|Red Wine Supernova                     |Chappell Roan    |    43|
|Black Planet                           |Sisters of Mercy |    33|
|Lucretia My Reflection - Vinyl Version |Sisters of Mercy |    33|
|Lullaby                                |The Cure         |    32|
|Pink Pony Club                         |Chappell Roan    |    32|
|All I Need                             |Radiohead        |    31|
|Here Comes Your Man                    |Pixies           |    31|
|Toxicity                               |System Of A Down |    31|
|Boys Don't Cry - Single Version        |The Cure         |    30|



Table: Top 5 artists of 2024

|Artist        | Plays|
|:-------------|-----:|
|Muse          |   406|
|Chappell Roan |   304|
|Ghost         |   283|
|One Direction |   210|
|The Cure      |   117|



Table: Top 10 songs of 2025

|Song                                            |Artist                      | Plays|
|:-----------------------------------------------|:---------------------------|-----:|
|Pumping Iron                                    |Andrew Lloyd Webber         |    29|
|Rolling Stock                                   |Andrew Lloyd Webber         |    24|
|Vanish Into You                                 |Lady Gaga                   |    24|
|Perfect Celebrity                               |Lady Gaga                   |    23|
|Marks Of The Evil One                           |Ghost                       |    21|
|Umbra                                           |Ghost                       |    21|
|Zombieboy                                       |Lady Gaga                   |    21|
|Cenotaph                                        |Ghost                       |    20|
|Crimson and Clover                              |Joan Jett & the Blackhearts |    20|
|Money For N



Table: Top 5 artists of 2025

|Artist              | Plays|
|:-------------------|-----:|
|Muse                |   344|
|Ghost               |   222|
|My Chemical Romance |   188|
|Lady Gaga           |   144|
|Chappell Roan       |   129|

## Lets discover how likely I am to skip certain songs. 

We'll be using the OneR package for machine learning, and we'll specifically use the streaming_history dataset. This dataset does not have songs under 20 seconds long removed, since it will otherwise assume I dont skip very many songs at all.

In [10]:
# mild cleaning, I dont want songs to be included that I haven't listened to at LEAST 10 times
skipped_songs <- streaming_history[streaming_history$song_title %in% names(which(table(streaming_history$song_title) >= 10)), ]

# ensure all variables are factors
skipped_songs[] <- lapply(skipped_songs, function(x) {
  if (is.character(x) | is.logical(x)) factor(x) else x
})

# removing columns that aren't neccesary
skipped_songs$ts = NULL
skipped_songs$ms_played = NULL
skipped_songs$track_url = NULL
skipped_songs$reason_start = NULL
skipped_songs$reason_end = NULL
skipped_songs$offline = NULL
skipped_songs$shuffle = NULL
skipped_songs$album_name = NULL

head(skipped_songs)

,song_title,artist_name,skipped
,<fct>,<fct>,<fct>
1,Gasoline,Halsey,TRUE
2,Gasoline,Halsey,TRUE
3,Gasoline,Halsey,TRUE
4,Gasoline,Halsey,TRUE
5,Gasoline,Halsey,TRUE
6,Semi-Automatic,Twenty One Pilots,TRUE


In [11]:
# install.packages("OneR")
library(OneR)

# using the OneR model to get my listening rules
skipped_songs_1R <- OneR(skipped ~ song_title, data = skipped_songs)
skipped_songs_1R



Call:
OneR.formula(formula = skipped ~ song_title, data = skipped_songs)

Rules:
If song_title = (Fuck A) Silver Lining                                                                          then skipped = FALSE
If song_title = (I Just) Died In Your Arms                                                                      then skipped = FALSE
If song_title = (One of Those) Crazy Girls                                                                      then skipped = FALSE
If song_title = 1-800-273-8255 (I Don't Wanna Be Alive)                                                         then skipped = FALSE
If song_title = 10:15 Saturday Night - 2005 Remaster                                                            then skipped = TRUE
If song_title = 11 Minutes (with Halsey feat. Travis Barker)                                                    then skipped = FALSE
If song_title = 1685 (feat. Muse)                                                                               then skip

In [12]:
skipped_songs_pred <- predict(skipped_songs_1R, skipped_songs)
table(actual = skipped_songs$skipped, predicted = skipped_songs_pred)

       predicted
actual  FALSE  TRUE
  FALSE 99185  7385
  TRUE  24867 14022

For songs, we have 77.83% accuracy

- [True Positive] correctly predicted skipped = 14,022
- [True Negative] correctly predicted NOT skipped = 99,185
- [False Positive] incorrectly predicted skipped = 7,385
- [False Negative] incorrectly predicted NOT skipped = 24,657

let's do the same process with artists

In [13]:
# mild cleaning, I dont want artists to be included that I haven't listened to at LEAST 3 times
skipped_artists <- streaming_history[streaming_history$artist_name %in% names(which(table(streaming_history$artist_name)>=3)), ]

skipped_artists[] <- lapply(skipped_artists, function(x) {
  if (is.character(x) | is.logical(x)) factor(x) else x
})

# removing collumns that aren't neccesary
skipped_artists$ts = NULL
skipped_artists$ms_played = NULL
skipped_artists$track_url = NULL
skipped_artists$reason_start = NULL
skipped_artists$reason_end = NULL
skipped_artists$offline = NULL
skipped_artists$shuffle = NULL
skipped_artists$album_name = NULL

head(skipped_artists)

,song_title,artist_name,skipped
,<fct>,<fct>,<fct>
1,Gasoline,Halsey,TRUE
2,Gasoline,Halsey,TRUE
3,Gasoline,Halsey,TRUE
4,Gasoline,Halsey,TRUE
5,Gasoline,Halsey,TRUE
6,Semi-Automatic,Twenty One Pilots,TRUE


In [14]:
skipped_artists_1R <- OneR(skipped ~ artist_name, data = skipped_artists)
skipped_artists_1R


Call:
OneR.formula(formula = skipped ~ artist_name, data = skipped_artists)

Rules:
If artist_name = 'Be More Chill' Ensemble                                                 then skipped = FALSE
If artist_name = *NSYNC                                                                   then skipped = FALSE
If artist_name = [bsd.u]                                                                  then skipped = FALSE
If artist_name = 3OH!3                                                                    then skipped = FALSE
If artist_name = 4 Door Theatre                                                           then skipped = FALSE
If artist_name = 5 Seconds of Summer                                                      then skipped = FALSE
If artist_name = 90´s or 20´s cartoons                                                    then skipped = FALSE
If artist_name = A-bug                                                                    then skipped = FALSE
If artist_name = a-ha      

In [15]:
skipped_artists_pred <- predict(skipped_artists_1R, skipped_artists)
table(actual = skipped_artists$skipped, predicted = skipped_artists_pred)

       predicted
actual   FALSE   TRUE
  FALSE 107182   6431
  TRUE   29771  11209

For artists, we have 76.58% accuracy

- [True Positive] correctly predicted skipped = 11,209
- [True Negative] correctly predicted NOT skipped = 107,182
- [False Positive] incorrectly predicted skipped = 6,431
- [False Negative] incorrectly predicted NOT skipped = 29,771

## What next?
The innaccuracies of this algorithm can likely be attributed to the size of the dataset. Most ML algorithms and libraries don't function nearly as well on larger datasets. 

I Initially planned to use another more robust algorithm such as Random Forest. Random Forest handles large skewed datasets far better than most algorithms, and the results would've allowed me to compare the results of OneR and Random Forest. However, upon training the model, it ran for 30 minutes and still had not finished, so, I ultimately had to make a sacrifice to meet the deadline of this project. 

Perhaps in the future I will run Random Forest on this data to compare the results.